src/logging.cpp

```
@@ -273,17 +318,45 @@ void BCLog::Logger::LogPrintStr(const std::string& str, const std::string& loggi
     if (m_print_to_file) {
         assert(m_fileout != nullptr);
 
-        // reopen the log file, if requested
-        if (m_reopen_file) {
-            m_reopen_file = false;
+        auto reopen_file = [&]() {
             FILE* new_fileout = fsbridge::fopen(m_file_path, "a");
             if (new_fileout) {
                 setbuf(new_fileout, nullptr); // unbuffered
                 fclose(m_fileout);
                 m_fileout = new_fileout;
+                m_file_size = 0;
             }
+        };
+
+        // reopen the log file, if requested
+        if (m_reopen_file) {
+            m_reopen_file = false;
+            reopen_file();
         }
         FileWriteStr(str_prefixed, m_fileout);
+        m_file_size += str_prefixed.size();
+
+        // If debug.log is large, rotate it (rename and recreate).
+        if (m_rotate_keep > 0 && m_file_size > m_file_limit * 1e6) {
+            // Rename debug.log to a name like debug-2021-06-25T22:57:54.log
+            int64_t seconds = GetTimeMicros()/1e6;
+            // Don't rotate within the same second to avoid name conflict.
+            if (m_last_rotate_time < seconds) {
+                m_last_rotate_time = seconds;
+                std::string now = FormatISO8601DateTime(seconds);
+                fs::path rotate{m_file_path};
+                rotate.remove_filename();
+                rotate /= "debug-" +  now + ".log";
+                try {
+                    fs::rename(m_file_path, rotate);
+                } catch (const fs::filesystem_error&) {};
```

We're ignoring any error from `rename`, not sure that's right.